### PR TITLE
Re-add features.sk

### PR DESCRIPTION
**Pull request by TheLimeGlass**
### Description
Re-add support for features.sk which was deleted potentially unintentionally. https://github.com/SkriptLang/Skript/pull/1604

---
Related Issues: https://github.com/SkriptLang/Skript/issues/3728

### CODE REVIEW

**Review by Moderocky — APPROVED** [review]

File: src/main/java/ch/njol/skript/Skript.java [hunk 1]
```diff
@@ -1408,20 +1424,23 @@ public static <E extends SkriptEvent> SkriptEventInfo<E> registerEvent(String na
...
 	 * 
 	 * @param name The name of the event, used for error messages
-	 * @param c The event's class
+	 * @param event The event's class
 	 * @param events The Bukkit events this event applies to
 	 * @param patterns Skript patterns to match this event
 	 * @return A SkriptEventInfo representing the registered event. Used to generate Skript's documentation.
 	 */
-	public static <E extends SkriptEvent> SkriptEventInfo<E> registerEvent(String name, Class<E> c, Class<? extends Event>[] events, String... patterns) {
+	public static <E extends SkriptEvent> SkriptEventInfo<E> registerEvent(String name, Class<E> event, Class<? extends Event>[] events, String... patterns) {
```
Comment (APickledWalrus) [inline src/main/java/ch/njol/skript/Skript.java:1432]: Do we also need to check `Skript.registerStructure`

File: src/main/java/ch/njol/skript/FeatureConfig.java [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+			SectionNode alter = (SectionNode) mc.getMainNode().get("alterSyntax");
+			if (alter != null && !alter.isEmpty()) {
+				for (Node node : new IteratorIterable<Node>(alter.iterator())) {	
+					alteredPatterns.put(node.getKey(), mc.get("alterSyntax", node.getKey()));
+				}
+			}
+		}
+	}
+
+	public static void discard() {
```
Comment (APickledWalrus) [inline src/main/java/ch/njol/skript/FeatureConfig.java:148]: should `alteredPatterns` also be cleared?

File: src/main/java/ch/njol/skript/FeatureConfig.java [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+					if (!af.exists())
+						saveTo = af;
+				} if (saveTo != null) {
+					InputStream in = zip.getInputStream(entry);
+					try {
+						assert in != null;
+						FileUtils.save(in, saveTo);
+					} finally {
+						in.close();
+					}
```
Comment (APickledWalrus) [inline src/main/java/ch/njol/skript/FeatureConfig.java:98]:
```suggestion
					try (InputStream in = zip.getInputStream(entry)) {
						FileUtils.save(in, saveTo);
					}
```
Should work

File: src/main/java/ch/njol/skript/FeatureConfig.java [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+		if (featureFile.exists()) {
+			try {
+				mc = new Config(featureFile, false, false, ":");
+			} catch (IOException e) {
+				e.printStackTrace();
+			}
+		} else {
+			ZipFile zip = null;
+			try {
+				zip = new ZipFile(file);
```
Comment (APickledWalrus) [inline src/main/java/ch/njol/skript/FeatureConfig.java:84]: Can also probably use try-with-resources here

File: src/main/java/ch/njol/skript/FeatureConfig.java [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+			mc.load(FeatureConfig.class);
+			// When we need to change values in the future of this file, bump the version node.
+			if (mc.get("version").isEmpty() || !mc.get("version").equalsIgnoreCase("1")) {
+				Skript.warning("Your features.sk config file is outdated. " +
+						"Backup any changes you've made, and delete your features.sk in the Skript folder to update it. " +
+						"After re-add any nodes you've changed.");
+			}
+			SectionNode section = (SectionNode) mc.getMainNode().get("features");
+			if (section != null && !section.isEmpty()) {
+				for (Node node : new IteratorIterable<Node>(section.iterator())) {
```
Comment (APickledWalrus) [inline src/main/java/ch/njol/skript/FeatureConfig.java:128]:
```suggestion
				for (Node node : section) {		
```

**Review by APickledWalrus — CHANGES_REQUESTED** [review]

**Review by AyhamAl-Ali — APPROVED** [review]

**Review by sovdeeth — APPROVED** [review]
Brilliant, this seems extremely useful, sad it got disabled.
I like the clear-er description changes, too.

File: src/main/java/ch/njol/skript/FeatureConfig.java [hunk 6]
```diff
@@ -0,0 +1,144 @@
...
+	public final static OptionSection alterSyntax = new OptionSection("alterSyntax");
+	public final static OptionSection features = new OptionSection("features");
+
+	private static Map<String, String> alteredPatterns = new HashMap<>();
+	private static List<String> disabledClassNames = new ArrayList<>();
+	private static List<String> disabledPatterns = new ArrayList<>();
+
+	private static boolean loaded;
+
+	public static Pair<Boolean, String[]> contains(String className, String... patterns) {
```
Comment (APickledWalrus) [inline src/main/java/ch/njol/skript/FeatureConfig.java:49]: Is there a reason for using a pair here? The second value doesn't seem to be used.

File: src/main/java/ch/njol/skript/FeatureConfig.java [hunk 7]
```diff
@@ -0,0 +1,144 @@
...
+		if (disabledClassNames.contains(className)) {
+			Skript.debug("Disabling feature '" + className + "' through the Features.sk config.");
+			return new Pair<Boolean, String[]>(true, patterns);
+		}
+
+		for (int i = 0 ; i < patterns.length ; i++) {
+			if (alteredPatterns.containsKey(patterns[i])) {
+				Skript.debug("Altering the pattern '" + patterns[i] + "' to '" + alteredPatterns.get(patterns[i]) + "'.");
+				patterns[i] = alteredPatterns.get(patterns[i]);
+				return new Pair<Boolean, String[]>(false, patterns);
```
Comment (APickledWalrus) [inline src/main/java/ch/njol/skript/FeatureConfig.java:59]: Is it possible that more than one pattern could be altered (for the same syntax)? This terminates after the first success.

File: src/main/java/ch/njol/skript/FeatureConfig.java [hunk 8]
```diff
@@ -0,0 +1,144 @@
...
+			} catch (IOException e) {
+				if (Skript.debug())
+					Skript.exception(e);
+			} finally {
+				featureFile = new File(Skript.getInstance().getDataFolder(), "features.sk");
+				try {
+					mc = new Config(featureFile, false, false, ":");
+				} catch (IOException e) {
+					e.printStackTrace();
+				}
```
Comment (APickledWalrus) [inline src/main/java/ch/njol/skript/FeatureConfig.java:102]: We could probably avoid this duplication by setting `loaded` later and calling this method again after the file is created (assuming no exception occurred).

File: src/main/java/ch/njol/skript/FeatureConfig.java [hunk 9]
```diff
@@ -0,0 +1,144 @@
...
+			}
+		}
+
+		if (mc != null) {
+			mc.load(FeatureConfig.class);
+			// When we need to change values in the future of this file, bump the version node.
+			if (mc.get("version").isEmpty() || !mc.get("version").equalsIgnoreCase("1")) {
+				Skript.warning("Your features.sk config file is outdated. " +
+						"Backup any changes you've made, and delete your features.sk in the Skript folder to update it. " +
+						"After re-add any nodes you've changed.");
```
Comment (APickledWalrus) [inline src/main/java/ch/njol/skript/FeatureConfig.java:112]:
```suggestion
						"After the updated file is created, re-add your changes.");
```

**Review by APickledWalrus — CHANGES_REQUESTED** [review]

**Comment by sovdeeth** [comment]
Closing due to inactivity

**Pull request closed without merging**